File: mRemoteV1/Controls/ListView.vb
Imports System.ComponentModel

Namespace Controls
    Public Class ListView
        Inherits System.Windows.Forms.ListView

#Region "Public Properties"

        <Category("Appearance"),
            DefaultValue(GetType(Color), "HighlightText")>
        Public Property HighlightForeColor As Color = SystemColors.HighlightText

        <Category("Appearance"),
            DefaultValue(GetType(Color), "Highlight")>
        Public Property HighlightBackColor As Color = SystemColors.Highlight

        <Category("Appearance"),
            DefaultValue(GetType(Color), "HotTrack")>
        Public Property HighlightBorderColor As Color = SystemColors.HotTrack

        <Category("Appearance"),
            DefaultValue(GetType(Color), "ControlText")>
        Public Property InactiveHighlightForeColor As Color = SystemColors.ControlText

        <Category("Appearance"),
            DefaultValue(GetType(Color), "Control")>
        Public Property InactiveHighlightBackColor As Color = SystemColors.Control

        <Category("Appearance"),
            DefaultValue(GetType(Color), "ControlDark")>
        Public Property InactiveHighlightBorderColor As Color = SystemColors.ControlDark

        <Category("Appearance"),
            DefaultValue(True)>
        Public Overloads Property ShowFocusCues As Boolean = True

        <Category("Appearance")>
        Public Property LabelAlignment As New Alignment(VerticalAlignment.Top, HorizontalAlignment.Left)

#End Region

#Region "Constructors"

        Public Sub New()
            OwnerDraw = True
        End Sub

#End Region

#Region "Protected Methods"

        Protected Overrides Sub OnDrawItem(e As DrawListViewItemEventArgs)
            If Not View = View.Tile Then MyBase.OnDrawItem(e)
            If e.ItemIndex < 0 Then MyBase.OnDrawItem(e)

            Dim foreColorBrush As Brush = Nothing
            Dim backColorBrush As Brush = Nothing
            Dim borderPen As Pen = Nothing
            Try
                If Focused Then
                    borderPen = New Pen(HighlightBorderColor)
                Else
                    borderPen = New Pen(InactiveHighlightBorderColor)
                End If

                If e.Item.Selected Then
                    If Focused Then
                        foreColorBrush = New SolidBrush(HighlightForeColor)
                        backColorBrush = New SolidBrush(HighlightBackColor)
                    Else
                        foreColorBrush = New SolidBrush(InactiveHighlightForeColor)
                        backColorBrush = New SolidBrush(InactiveHighlightBackColor)
                    End If
                Else
                    foreColorBrush = New SolidBrush(e.Item.ForeColor)
                    backColorBrush = New SolidBrush(BackColor)
                End If

                e.Graphics.FillRectangle(backColorBrush, e.Bounds)

                If Focused And ShowFocusCues Then
                    e.DrawFocusRectangle()
                ElseIf e.Item.Selected Then
                    e.Graphics.DrawRectangle(borderPen, e.Bounds.X, e.Bounds.Y, e.Bounds.Width - 1, e.Bounds.Height - 1)
                End If

                Dim imageBounds As New Rectangle(e.Bounds.X + 2, e.Bounds.Y + 6, 16, 16)
                Dim textBounds As Rectangle = e.Bounds

                If e.Item.ImageList IsNot Nothing Then
                    Dim image As Image = Nothing
                    If _
                        Not String.IsNullOrEmpty(e.Item.ImageKey) And
                        e.Item.ImageList.Images.ContainsKey(e.Item.ImageKey) Then
                        image = e.Item.ImageList.Images.Item(e.Item.ImageKey)
                    ElseIf Not e.Item.ImageIndex < 0 And e.Item.ImageList.Images.Count > e.Item.ImageIndex Then
                        image = e.Item.ImageList.Images(e.Item.ImageIndex)
                    End If
                    If image IsNot Nothing Then
                        e.Graphics.DrawImageUnscaledAndClipped(image, imageBounds)
                        textBounds.X = textBounds.Left + 20
                        textBounds.Width = textBounds.Width - 20
                    End If
                End If

                e.Graphics.DrawString(e.Item.Text, e.Item.Font, foreColorBrush, textBounds, GetStringFormat())
            Finally
                If foreColorBrush IsNot Nothing Then foreColorBrush.Dispose()
                If backColorBrush IsNot Nothing Then backColorBrush.Dispose()
                If borderPen IsNot Nothing Then borderPen.Dispose()
            End Try
        End Sub

#End Region

#Region "Private Methods"

        Private Function GetStringFormat() As StringFormat
            Dim format As StringFormat = StringFormat.GenericDefault

            Select Case LabelAlignment.Vertical
                Case VerticalAlignment.Top
                    format.LineAlignment = StringAlignment.Near
                Case VerticalAlignment.Middle
                    format.LineAlignment = StringAlignment.Center
                Case VerticalAlignment.Bottom
                    format.LineAlignment = StringAlignment.Far
            End Select

            Select Case LabelAlignment.Horizontal
                Case HorizontalAlignment.Left
                    format.Alignment = StringAlignment.Near
                Case HorizontalAlignment.Center
                    format.Alignment = StringAlignment.Center
                Case HorizontalAlignment.Right
                    format.Alignment = StringAlignment.Far
            End Select

            If RightToLeft Then
                format.FormatFlags = format.FormatFlags Or StringFormatFlags.DirectionRightToLeft
            End If

            If LabelWrap Then
                format.FormatFlags = format.FormatFlags And Not StringFormatFlags.NoWrap
            Else
                format.FormatFlags = format.FormatFlags Or StringFormatFlags.NoWrap
            End If

            Return format
        End Function

#End Region
    End Class

    <TypeConverter(GetType(ExpandableObjectConverter))>
    Public Class Alignment
        Public Sub New()
        End Sub

        Public Sub New(verticalAlignment As VerticalAlignment, horizontalAlignment As HorizontalAlignment)
            Vertical = verticalAlignment
            Horizontal = horizontalAlignment
        End Sub

        <NotifyParentProperty(True),
            DefaultValue(VerticalAlignment.Top)>
        Public Property Vertical As VerticalAlignment = VerticalAlignment.Top

        <NotifyParentProperty(True),
            DefaultValue(HorizontalAlignment.Left)>
        Public Property Horizontal As HorizontalAlignment = HorizontalAlignment.Left

        Public Overrides Function ToString() As String
            Return String.Format("{0}, {1}", Vertical, Horizontal)
        End Function
    End Class

    Public Enum VerticalAlignment
        Top
        Middle
        Bottom
    End Enum

    Public Enum HorizontalAlignment
        Left
        Center
        Right
    End Enum
End Namespace

File: mRemoteV1.ConvertedToC#/Controls/ListView.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.ComponentModel;

namespace mRemoteNG.Controls
{
	public class ListView : Windows.Forms.ListView
	{

		#region "Public Properties"
		[Category("Appearance"), DefaultValue(typeof(Color), "HighlightText")]
		public Color HighlightForeColor { get; set; }

		[Category("Appearance"), DefaultValue(typeof(Color), "Highlight")]
		public Color HighlightBackColor { get; set; }

		[Category("Appearance"), DefaultValue(typeof(Color), "HotTrack")]
		public Color HighlightBorderColor { get; set; }

		[Category("Appearance"), DefaultValue(typeof(Color), "ControlText")]
		public Color InactiveHighlightForeColor { get; set; }

		[Category("Appearance"), DefaultValue(typeof(Color), "Control")]
		public Color InactiveHighlightBackColor { get; set; }

		[Category("Appearance"), DefaultValue(typeof(Color), "ControlDark")]
		public Color InactiveHighlightBorderColor { get; set; }

		[Category("Appearance"), DefaultValue(true)]
		public bool ShowFocusCues { get; set; }

		[Category("Appearance")]
		public Alignment LabelAlignment { get; set; }
		#endregion

		#region "Constructors"
		public ListView()
		{
			OwnerDraw = true;
		}
		#endregion

		#region "Protected Methods"
		protected override void OnDrawItem(DrawListViewItemEventArgs e)
		{
			if (!(View == View.Tile))
				base.OnDrawItem(e);
			if (e.ItemIndex < 0)
				base.OnDrawItem(e);

			Brush foreColorBrush = null;
			Brush backColorBrush = null;
			Pen borderPen = null;
			try {
				if (Focused) {
					borderPen = new Pen(HighlightBorderColor);
				} else {
					borderPen = new Pen(InactiveHighlightBorderColor);
				}

				if (e.Item.Selected) {
					if (Focused) {
						foreColorBrush = new SolidBrush(HighlightForeColor);
						backColorBrush = new SolidBrush(HighlightBackColor);
					} else {
						foreColorBrush = new SolidBrush(InactiveHighlightForeColor);
						backColorBrush = new SolidBrush(InactiveHighlightBackColor);
					}
				} else {
					foreColorBrush = new SolidBrush(e.Item.ForeColor);
					backColorBrush = new SolidBrush(BackColor);
				}

				e.Graphics.FillRectangle(backColorBrush, e.Bounds);

				if (Focused & ShowFocusCues) {
					e.DrawFocusRectangle();
				} else if (e.Item.Selected) {
					e.Graphics.DrawRectangle(borderPen, e.Bounds.X, e.Bounds.Y, e.Bounds.Width - 1, e.Bounds.Height - 1);
				}

				Rectangle imageBounds = new Rectangle(e.Bounds.X + 2, e.Bounds.Y + 6, 16, 16);
				Rectangle textBounds = e.Bounds;

				if (e.Item.ImageList != null) {
					Image image = null;
					if (!string.IsNullOrEmpty(e.Item.ImageKey) & e.Item.ImageList.Images.ContainsKey(e.Item.ImageKey)) {
						image = e.Item.ImageList.Images[e.Item.ImageKey];
					} else if (!(e.Item.ImageIndex < 0) & e.Item.ImageList.Images.Count > e.Item.ImageIndex) {
						image = e.Item.ImageList.Images[e.Item.ImageIndex];
					}
					if (image != null) {
						e.Graphics.DrawImageUnscaledAndClipped(image, imageBounds);
						textBounds.X = textBounds.Left + 20;
						textBounds.Width = textBounds.Width - 20;
					}
				}

				e.Graphics.DrawString(e.Item.Text, e.Item.Font, foreColorBrush, textBounds, GetStringFormat());
			} finally {
				if (foreColorBrush != null)
					foreColorBrush.Dispose();
				if (backColorBrush != null)
					backColorBrush.Dispose();
				if (borderPen != null)
					borderPen.Dispose();
			}
		}
		#endregion

		#region "Private Methods"
		private StringFormat GetStringFormat()
		{
			StringFormat format = StringFormat.GenericDefault;

			switch (LabelAlignment.Vertical) {
				case VerticalAlignment.Top:
					format.LineAlignment = StringAlignment.Near;
					break;
				case VerticalAlignment.Middle:
					format.LineAlignment = StringAlignment.Center;
					break;
				case VerticalAlignment.Bottom:
					format.LineAlignment = StringAlignment.Far;
					break;
			}

			switch (LabelAlignment.Horizontal) {
				case HorizontalAlignment.Left:
					format.Alignment = StringAlignment.Near;
					break;
				case HorizontalAlignment.Center:
					format.Alignment = StringAlignment.Center;
					break;
				case HorizontalAlignment.Right:
					format.Alignment = StringAlignment.Far;
					break;
			}

			if (RightToLeft) {
				format.FormatFlags = format.FormatFlags | StringFormatFlags.DirectionRightToLeft;
			}

			if (LabelWrap) {
				format.FormatFlags = format.FormatFlags & !StringFormatFlags.NoWrap;
			} else {
				format.FormatFlags = format.FormatFlags | StringFormatFlags.NoWrap;
			}

			return format;
		}
		#endregion
	}

	[TypeConverter(typeof(ExpandableObjectConverter))]
	public class Alignment
	{

		public Alignment()
		{
		}

		public Alignment(VerticalAlignment verticalAlignment, HorizontalAlignment horizontalAlignment)
		{
			Vertical = verticalAlignment;
			Horizontal = horizontalAlignment;
		}

		[NotifyParentProperty(true), DefaultValue(VerticalAlignment.Top)]
		public VerticalAlignment Vertical { get; set; }

		[NotifyParentProperty(true), DefaultValue(HorizontalAlignment.Left)]
		public HorizontalAlignment Horizontal { get; set; }

		public override string ToString()
		{
			return string.Format("{0}, {1}", Vertical, Horizontal);
		}
	}

	public enum VerticalAlignment : int
	{
		Top,
		Middle,
		Bottom
	}

	public enum HorizontalAlignment : int
	{
		Left,
		Center,
		Right
	}
}
